File: Documents/GitHub/Web-Projects/project2/input_validator.rb
# Constants
TOKEN_PATTERN = /
  \d*\.\d+ |   # Decimal numbers
  \d+ |        # Integer numbers
  [\+\-\*\/\%] | # Operators
  \(|\)|       # Parentheses
  √|∛|log|,|sin|cos|tan|exponent|percentage|factorial|temp|odds|evens|squares|primes|is_prime|q|π # Functions
/x
# Responsible for checking the input before calculating
module InputValidator

  # Validates if the given input string is a properly formatted mathematical expression
  # @param s [String] The input string representing a mathematical expression
  # @return [Boolean] Returns true if the input string is valid, false otherwise
  def valid_input?(s)

    #Check for valid operator placement
    return false unless valid_operators?(s)

    #check for valid paretnheses placement
    return false unless parentheses_balanced?(s)

    true
  end

  # Method to check for balanced parentheses
  # @param s [String] The input string to check for balanced parentheses
  # @return [Boolean] Returns true if the parentheses are balanced, false otherwise
  def parentheses_balanced?(s)
    stack = []
    s.each_char do |char|
      if char == '('
        stack.push(char)
      elsif char == ')'
        return false if stack.empty? || stack.pop != '('
      end
    end
    stack.empty? # If stack is empty, parentheses are balanced
  end

  # Validates the placement of operators in the given string to ensure correct syntax
  # @param s [String] The input string containing the expression to validate
  # @return [Boolean] Returns true if the operators are placed correctly, false otherwise
  def valid_operators?(s)
    tokens = s.scan(TOKEN_PATTERN)
    
    # Debug: Print tokens for inspection
    puts "Debug: Tokens: #{tokens.inspect}"

    # Check if tokens are empty or start/end with invalid tokens
    return false if tokens.empty?
    return false if tokens.first =~ /[\+\-\*\/\%\)]|[ \t]/ 
    return false if tokens.last =~ /[\+\-\*\/\%\(]|[ \t,]/

    tokens.each_cons(2) do |prev, curr|
      # Debug: Print previous and current tokens
      puts "Checking tokens: prev=#{prev}, curr=#{curr}"

      # Check that operators aren't next to each other
      if prev =~ /[\+\-\*\/\%]/ && curr =~ /[\+\-\*\/\%]/
        puts "Consecutive operators: #{prev} #{curr}"
        return false
      end

      # Check no operator is directly after an opening parenthesis or before a closing parenthesis
      if prev == '(' && curr =~ /[\+\-\*\/\%]/ 
        puts "Operator after '(': #{prev} #{curr}"
        return false
      end
      if curr == ')' && prev =~ /[\+\-\*\/\%]/
        puts "Operator before ')': #{prev} #{curr}"
        return false
      end

      if prev =~ /\d*\.\d+|\d+/ && curr =~ /[(|√|∛|log|sin|cos|tan|exponent|percentage|factorial|temp|odds|evens|squares|primes|is_prime|π]/
        puts "digit before '(' or function: #{prev} #{curr}"
        return false
      end
      if prev == ')' && curr =~ /\d*\.\d+|\d+/
        puts "digit after ')' or : #{prev} #{curr}"
        return false
      end

      # Ensure functions are followed by an opening parenthesis
      if prev =~ /sin|cos|tan|log|percentage|factorial|temp|absolute|isPrime|exponent|odds|evens|is_prime|squares|primes|√|∛/ && curr != '('
        puts "Function without '(': #{prev} #{curr}"
        return false
      end

      # Ensure "percentage", "exponent", etc., are followed by a valid structure
      if (prev == 'percentage' || prev == 'exponent' || prev == 'absolute' || prev == 'odds' || prev == 'evens' || prev == 'squares') && (curr != '(' || !s.include?(','))
        puts "Invalid structure for: #{prev} #{curr}"
        return false
      end
    end

    true
  end
end

File: Documents/GitHub/Web-Projects/project5/input_validator.js
const TOKEN_PATTERN = /\d*\.\d+|\d+|[\+\-\*\/\%]|\(|\)|√|∛|log|,|sin|cos|tan|exponent|percentage|factorial|temp|odds|evens|squares|mean|mode|median|fine_smallest|primes|is_prime|neg|π/g;
// checks input for valid syntex
class InputValidator {

    // main valid checker
    is_input_valid(s) {
        return this.valid_operators(s) && this.parentheses_balanced(s);
    }

    // check for () syntax
    parentheses_balanced(s) {
        const stack = [];
        for(const char of s) {
            if(char === '(') {
            stack.push(char);
        } else if (char === ')') {
            if(stack.length === 0 || stack.pop() !== '(') {
                return false;
            }
        }
    }
    return stack.length === 0;
}
    
    // check for operator syntax
    valid_operators(s) {
        const tokens = s.match(TOKEN_PATTERN) || [];

        console.log("Tokens:", tokens);

        if (tokens.length === 0) return false;
        if (/[\+\-\*\/\%\)]|[ \t]/.test(tokens[0])) return false;
        if (/[\+\-\*\/\%\(]|[ \t,]/.test(tokens[tokens.length - 1])) return false;

        for(let i = 0; i < tokens.length - 1; i++) {
            const prev = tokens[i];
            const curr = tokens[i+1];
            console.log("Checcking tokens:", prev, curr);

            if (/[\+\-\*\/\%]/.test(prev) && /[\+\-\*\/\%]/.test(curr)) {
                console.log("Consecutive operators:", prev, curr);
                return false;
            }
            
            if (prev === '(' && /[\+\-\*\/\%]/.test(curr)) {
                console.log("Operator after '(':", prev, curr);
                return false;
            }
    
            if (curr === ')' && /[\+\-\*\/\%]/.test(prev)) {
                console.log("Operator before ')':", prev, curr);
                return false;
            }
    
            if (/\d*\.\d+|\d+/.test(prev) && /[(|√|∛|log|sin|cos|tan|exponent|percentage|factorial|temp|odds|evens|mean|squares|primes|is_prime|π]/.test(curr)) {
                console.log("Digit before '(' or function:", prev, curr);
                return false;
            }
    
            if (prev === ')' && /\d*\.\d+|\d+/.test(curr)) {
                console.log("Digit after ')':", prev, curr);
                return false;
            }
    
            if (/sin|cos|tan|log|percentage|factorial|temp|absolute|is_prime|exponent|odds|mean|evens|is_prime|squares|primes|√|∛/.test(prev) && curr !== '(') {
                console.log("Function without '(':", prev, curr);
                return false;
            }
    
            if (['percentage', 'exponent', 'absolute', 'odds', 'evens', 'squares'].includes(prev) && (curr !== '(' || !s.includes(','))) {
                console.log("Invalid structure for:", prev, curr);
                return false;
            }
        }
    return true;
    }

}

export default InputValidator;